internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,479 @@
+func makeResponse(resp *responseInfo) *SinglePingResponse {
+	var data *model.ArchivalMaybeBinaryData
+	if resp.raw != nil {
+		data = &model.ArchivalMaybeBinaryData{Value: string(resp.raw)}
+	}
+	return &SinglePingResponse{
+		Data:              data,
+		Failure:           archival.NewFailure(resp.err),
+		ResponseTime:      formatTime(resp.t),
+		SupportedVersions: resp.versions,
+	}
+}
+
+// Measurer performs the measurement.
+type Measurer struct {
+	config Config
+}
+
+// ExperimentName implements ExperimentMeasurer.ExperimentName.
+func (m *Measurer) ExperimentName() string {
+	return testName
+}
+
+// ExperimentVersion implements ExperimentMeasurer.ExperimentVersion.
+func (m *Measurer) ExperimentVersion() string {
+	return testVersion
+}
+
+// pingInfo contains information about a ping request
+// and the corresponding ping responses
+type pingInfo struct {
+	request   *requestInfo
+	responses []*responseInfo
+}
+
+// requestInfo contains the information of a sent ping request.
+type requestInfo struct {
+	t     time.Time
+	raw   []byte
+	dstID string
+	srcID string
+	err   error
+}
+
+// responseInfo contains the information of a received ping reponse.
+type responseInfo struct {
+	t        time.Time
+	raw      []byte
+	dstID    string
+	versions []uint32
+	err      error
+}
+
+// sender sends a ping requests to the target hosts every second
+func (m *Measurer) sender(
+	ctx context.Context,
+	pconn model.UDPLikeConn,
+	destAddr *net.UDPAddr,
+	out chan<- requestInfo,
+	sess model.ExperimentSession,
+) {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	i := 0
+	for i < int(m.config.repetitions()) {
+		select {
+		case <-ctx.Done():
+			return // user aborted or timeout expired
+
+		case sendTime := <-ticker.C:
+			packet, dstID, srcID := buildPacket()     // build QUIC Initial packet
+			_, err := pconn.WriteTo(packet, destAddr) // send Initial packet
+
+			sess.Logger().Infof("PING %s", destAddr)
+
+			// propagate send information, including errors
+			out <- requestInfo{raw: packet, t: sendTime, dstID: hex.EncodeToString(dstID), srcID: hex.EncodeToString(srcID), err: err}
+			i += 1
+		}
+	}
+}
+
+// receiver receives incoming server responses and
+// dissects the payload of the version negotiation response
+func (m *Measurer) receiver(
+	ctx context.Context,
+	pconn model.UDPLikeConn,
+	out chan<- responseInfo,
+	sess model.ExperimentSession,
+) {
+	for ctx.Err() == nil {
+		// read (timeout was set in Run)
+		buffer := make([]byte, 1024)
+		n, addr, err := pconn.ReadFrom(buffer)
+		respTime := time.Now()
+		if err != nil {
+			// stop if the connection is already closed
+			if errors.Is(err, net.ErrClosed) {
+				break
+			}
+			// store read failures and continue receiving
+			out <- responseInfo{t: respTime, err: err}
+			continue
+		}
+		resp := buffer[:n]
+
+		// dissect server response
+		supportedVersions, dst, err := m.DissectVersionNegotiation(resp)
```

Comment:
```suggestion
		supportedVersions, dst, err := m.dissectVersionNegotiation(resp)
```

This method should be private